torch_xla/csrc/status.cpp

```diff
@@ -38,38 +72,43 @@ absl::Status MaybeWithNewMessage(const absl::Status& status, const char* file,
     return status;
   }
 
-  std::string_view old_message = status.message();
-
   // Replace the old status message with `new_message`, if it's not empty.
   //
   // The idea is that whenever `new_message` is given, it should have more
   // context to give a better error message to the user.
-  std::string_view message = new_message.empty() ? old_message : new_message;
+  auto new_status = absl::Status(
+      status.code(), new_message.empty() ? status.message() : new_message);
 
-  // If `TORCH_SHOW_CPP_STACKTRACES` is set, show the context of this error.
-  // In other words, show:
-  //   1. The error location
-  //   2. The old messages that were replaced by `new_message`.
+  // If `TORCH_SHOW_CPP_STACKTRACES` is set:
   //
-  // This should give more context for developers. Showing the older error
-  // messages alongside their debug information.
+  //     1. append the current stack frame to the status propagation trace
+  //        payload
   //
-  // Note that we also condition showing source location information by (2)
-  // (i.e. `new_message` is not empty) because we don't really wish to show
-  // a stacktrace. Instead, we show only the history of error messages that
-  // has led to the current error.
-  const std::string context =
-      (torch::get_cpp_stacktraces_enabled() && !new_message.empty())
-          ? absl::StrCat(LocationStrWithSpace(file, line),
-                         "\nFrom Error: ", old_message)
-          : "";
-
-  return absl::Status(status.code(), absl::StrCat(message, context));
+  //     2. append the new error message, if not empty
+  if (torch::get_cpp_stacktraces_enabled()) {
+    auto status_propagation_trace = GetStatusPropagationTraceOrEmpty(status);
+    status_propagation_trace.Append(
+        GetStackFrame(file, line, function, new_message));
+    new_status.SetPayload(kStatusPropagationTraceKey, status_propagation_trace);
+  }
+
+  return new_status;
 }
 
 void MaybeThrow(const absl::Status& status) {
   if (!status.ok()) {
-    throw std::runtime_error(std::string(status.message()));
+    auto status_propagation_trace = GetStatusPropagationTraceOrEmpty(status);
+    auto status_propagation_trace_str =
+        status_propagation_trace.empty()
+            ? ""
+            : absl::StrCat("\n\nStatus Propagation Trace:",
+                           status_propagation_trace.Flatten());
+    auto cpp_stacktrace_str =
+        torch::get_cpp_stacktraces_enabled()
+            ? absl::StrCat("\n\nC++ Stacktrace:\n", tsl::CurrentStackTrace())
+            : "";
+    throw std::runtime_error(absl::StrCat(
```

Comment:
When TORCH_SHOW_CPP_STACKTRACES is set, TORCH_CHECK would automatically include the stack trace. Therefore if we replace this with TORCH_CHECK(), we can remove the logic for capturing the stacktrace ourselves.